pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,498 @@
+	}
+
+	//TODO in a webhook annotate the broker object with its ingress details
+	// Route to shared ingress with namespace/name in the path as the broker
+	// identifier.
+	b.Status.SetAddress(&apis.URL{
+		Scheme: "http",
+		Host:   names.ServiceHostName("broker-ingress", system.Namespace()),
+		Path:   fmt.Sprintf("/%s/%s", b.Namespace, b.Name),
+	})
+
+	// Verify the ingress service is healthy via endpoints.
+	ingressEndpoints, err := r.endpointsLister.Endpoints(system.Namespace()).Get(ingressServiceName)
+	if err != nil {
+		logger.Error("Problem getting endpoints for ingress", zap.String("namespace", system.Namespace()), zap.Error(err))
+		b.Status.MarkIngressFailed("ServiceFailure", "%v", err)
+		return err
+	}
+	b.Status.PropagateIngressAvailability(ingressEndpoints)
+
+	r.targetsConfig.MutateBroker(b.Namespace, b.Name, func(m config.BrokerMutation) {
+		m.SetID(string(b.UID))
+		m.SetAddress(b.Status.Address.URL.String())
+		m.SetDecoupleQueue(&config.Queue{
+			//TODO should this use the status fields or the name generator funcs?
+			Topic:        b.Status.TopicID,
+			Subscription: b.Status.SubscriptionID,
+		})
+		if b.Status.IsReady() {
+			m.SetState(config.State_READY)
+		} else {
+			m.SetState(config.State_UNKNOWN)
+		}
+	})
+
+	// Update config map
+	// TODO should this happen in a defer so there's an update regardless of
+	// error status, or only when reconcile is successful?
+	r.flagTargetsForUpdate()
+	return nil
+}
+
+func (r *Reconciler) reconcileDecouplingTopicAndSubscription(ctx context.Context, b *brokerv1beta1.Broker) error {
+	logger := logging.FromContext(ctx)
+	logger.Debug("Reconciling decoupling topic", zap.Any("broker", b))
+	// get ProjectID from metadata if projectID isn't set
+	projectID, err := utils.ProjectID(r.projectID)
+	if err != nil {
+		logger.Error("Failed to find project id", zap.Error(err))
+		return err
+	}
+	// Set the projectID in the status.
+	b.Status.ProjectID = projectID
+
+	client, err := r.CreateClientFn(ctx, projectID)
+	if err != nil {
+		logger.Error("Failed to create Pub/Sub client", zap.Error(err))
+		return err
+	}
+	defer client.Close()
+
+	// Check if topic exists, and if not, create it.
+	topicID := resources.GenerateDecouplingTopicName(b)
+	topic := client.Topic(topicID)
+	exists, err := topic.Exists(ctx)
+	if err != nil {
+		logger.Error("Failed to verify Pub/Sub topic exists", zap.Error(err))
+		return err
+	}
+
+	if !exists {
+		topicConfig := &gpubsub.TopicConfig{
+			Labels: map[string]string{
+				"resource":     "brokers",
+				"broker_class": brokerv1beta1.BrokerClass,
+				"namespace":    b.Namespace,
+				"name":         b.Name,
+				//TODO add resource labels, but need to be sanitized: https://cloud.google.com/pubsub/docs/labels#requirements
+			},
+		}
+		// Create a new topic.
+		logger.Debug("Creating topic with cfg", zap.String("id", topicID), zap.Any("cfg", topicConfig))
+		topic, err = client.CreateTopicWithConfig(ctx, topicID, topicConfig)
+		if err != nil {
```

Comment:
Maybe check for AlreadyExists here to guard against race conditions?

Reply:
I don't think that's necessary There shouldn't be any other controller threads operating on this broker. If for some reason something else creates a topic with the same name, the reconcile will fail, retry, and skip this step.